fleet/postgres/agent_groups.go

```diff
@@ -28,11 +30,163 @@ type agentGroupRepository struct {
+		from
+			(select
+				ag.id,
+				ag.name,
+				ag.description,
+				ag.mf_owner_id,
+				ag.mf_channel_id,
+				ag.tags,
+				ag.ts_created,
+				sum(case when agm.agent_groups_id is not null then 1 else 0 end) as total,
+				sum(case when agm.agent_state = 'online' then 1 else 0 end) as online
+			from agent_groups ag
+				left join agent_group_membership agm
+					on ag.id = agm.agent_groups_id
+					and ag.mf_owner_id = agm.mf_owner_id
+			WHERE ag.mf_owner_id = :mf_owner_id %s
+				group by ag.id,
+					ag.name,
+					ag.description,
+					ag.mf_owner_id,
+					ag.mf_channel_id,
+					ag.tags,
+					ag.ts_created)
+			as agent_groups %s%s ORDER BY %s %s LIMIT :limit OFFSET :offset;`, nameQuery, tagsQuery, metadataQuery, orderQuery, dirQuery)
+
+	params := map[string]interface{}{
+		"mf_owner_id": ownerID,
+		"limit":       pm.Limit,
+		"offset":      pm.Offset,
+		"name":        name,
+		"metadata":    metadata,
+		"tags":        tags,
+	}
+	rows, err := a.db.NamedQueryContext(ctx, q, params)
+	if err != nil {
+		return fleet.PageAgentGroup{}, errors.Wrap(errors.ErrSelectEntity, err)
+	}
+	defer rows.Close()
+
+	var items []fleet.AgentGroup
+	for rows.Next() {
+		dbAgentGroup := dbAgentGroup{MFOwnerID: ownerID}
+		if err := rows.StructScan(&dbAgentGroup); err != nil {
+			return fleet.PageAgentGroup{}, errors.Wrap(errors.ErrSelectEntity, err)
+		}
+
+		agentGroup, err := toAgentGroup(dbAgentGroup)
+		if err != nil {
+			return fleet.PageAgentGroup{}, errors.Wrap(errors.ErrSelectEntity, err)
+		}
+
+		items = append(items, agentGroup)
+	}
+
+	count := fmt.Sprintf(
+		`select
+			COUNT(*)
+		from
+		(select
+			ag.id,
+			ag.name,
+			ag.description,
+			ag.mf_owner_id,
+			ag.mf_channel_id,
+			ag.tags,
+			ag.ts_created,
+			sum(case when agm.agent_groups_id is not null then 1 else 0 end) as total,
+			sum(case when agm.agent_state = 'online' then 1 else 0 end) as online
+		from agent_groups ag
+			left join agent_group_membership agm
+				on ag.id = agm.agent_groups_id
+					and ag.mf_owner_id = agm.mf_owner_id
+		WHERE ag.mf_owner_id = :mf_owner_id %s
+		group by ag.id,
+			ag.name,
+			ag.description,
+			ag.mf_owner_id,
+			ag.mf_channel_id,
+			ag.tags,
+			ag.ts_created) 
+		as agent_groups %s%s;`, nameQuery, tagsQuery, metadataQuery)
+
+	total, err := total(ctx, a.db, count, params)
+	if err != nil {
+		return fleet.PageAgentGroup{}, errors.Wrap(errors.ErrSelectEntity, err)
+	}
+
+	page := fleet.PageAgentGroup{
+		AgentGroups: items,
+		PageMetadata: fleet.PageMetadata{
+			Total:  total,
+			Offset: pm.Offset,
+			Limit:  pm.Limit,
+			Order:  pm.Order,
+			Dir:    pm.Dir,
+		},
+	}
+
+	return page, nil
 }
 
 func (a agentGroupRepository) RetrieveByID(ctx context.Context, groupID string, ownerID string) (fleet.AgentGroup, error) {
-	q := `SELECT id, name, mf_owner_id, mf_channel_id, tags, ts_created FROM agent_groups WHERE id = $1 AND mf_owner_id = $2`
+	q :=
```

Comment:
I was originally thinking we would do two queries - the select on agent_groups and then a separate SELECT COUNT on agent_group_membership for the total/online but if this works and you think it's efficient, that's fine

Reply:
When we have more data, i can check if this query its using too much cpu with scan tables and i can optimize with some new indexes, or refactor to two separate queries.